APE13.rst

```diff
@@ -0,0 +1,450 @@
+new type names, which the Astropy Project will facilitate as needed. If a
+``custom:`` type name is needed, we recommend that these be coordinated and
+agreed as much as possible between different packages to make sure that these
+can be useful (which would not be the case if each package created their own
+set of custom type names).
+
+High-level Astropy Object
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Unlike the low-level API, the ‘high-level’ interface described here will be a
+single Astropy-developed class since it interfaces with various Astropy objects.
+This high-level API would provide the ability for example to get ``SkyCoord``,
+``Time`` etc. objects back from a pixel to world conversion, and conversely to
+be able to convert ``SkyCoord``, ``Time`` etc. to pixel values.
+
+The high-level object would not inherit from the low-level classes but instead
+wrap them. The high-level object should provide at a minimum the
+following two methods:
+
+.. code-block:: python
+
+    def pixel_to_world(self, *pixel_arrays):
+        """
+        Convert pixel coordinates to world coordinates (represented by Astropy
+        objects).
+        """
+
+    def world_to_pixel(self, *world_objects):
+        """
+        Convert world coordinates (represented by Astropy objects) to pixel
+        coordinates
+        """
+
+The low-level object must be available under the attribute name ``low_level_wcs``
+and the low-level methods such as ``pixel_to_world_values`` will thus be
+available by doing:
+
+.. code-block:: python
+
+    >>> wcs.low_level_wcs.pixel_to_world_values(...)
+
+Since a single Astropy object might correspond to two non-contiguous dimensions
+in the WCS (for example the first and third world dimensions), we need to
+specify the rules for the order in which Astropy objects are returned from the
+high-level ``pixel_to_world`` method, and in which order they should be given to
+the high-level ``world_to_pixel`` method. The standard order should be that
+given by considering only the first occurrence of the coordinate alias string in
+``world_axis_object_components``. For example, if
+``world_axis_object_components`` is
+
+.. code-block:: python
+
+    [('skycoord', 'ra'), ('time', 0), ('skycoord', 'dec')]
+
+Then the order of the Astropy objects should be ``SkyCoord`` then ``Time`` (we
+essentially ignore ``('skycoord', 'dec')``). This rule will always be followed
+for ``pixel_to_world``, but on the other hand provided there is no ambiguity,
+``world_to_pixel`` could be more forgiving if the coordinates are specified in
+the wrong order (though an error should be raised if there are any ambiguities
+and the order is not the standard one).
+
+Branches and pull requests
+--------------------------
+
+N/A
+
+Implementation
+--------------
+
+The low-level API could be implemented specifically for FITS-WCS into the
+Astropy core package, along with the more general high-level Astropy object.
```

Comment:
Could the low-level API be added to the existing `astropy.wcs.WCS` class? Or probably a new class wrapping it would be needed? I.e. then one would end up with this, no?
```
wcs_high_level.wcs_low_level.wcs_legacy
```
Suggest to clarify that point.

Reply:
I think we didn't really decide in this APE how this will be done for astropy.wcs specifically, we left it pretty general. Since we have control over the astropy.wcs code we could easily just add this interface to the existing WCS object to avoid too many nested levels.